### PR TITLE
chore(TDKN-239): Add Digester to crypto-utils

### PR DESCRIPTION
**Pull request by fhuaulme**
**What is the problem this Pull Request is trying to solve?**
 
Provide a way to digest a value using a salt.

**What is the chosen solution to this problem?**
 
Digest a value using "(salt, delimiter, value) -> (salt + delimiter + sha256(value))" as result (instead of a single `value -> sha256(value)`).

**Link to the JIRA issue**

https://jira.talendforge.org/browse/TDKN-239
 
**Please check if the Pull Request fulfills these requirements**
- [ ] The PR commit message follows our [guidelines](https://github.com/Talend/daikon/blob/master/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features, coverage should be over 75% in the new code)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in JIRA), if any, are all linked or available in the Pull Request

 
**[ ] This Pull Request introduces a breaking change**

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/13/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/13/)

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+public class Digester {
+
+    private final DigestSource digestSource;
+
+    private final KeySource keySource;
+
+    private final String delimiter;
+
+    public Digester(KeySource keySource, String delimiter) {
+        this.keySource = keySource;
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: Suggest adding another constructor that doesn't take a delimiter, and default it to something sensible.

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        this.delimiter = delimiter;
+        this.digestSource = DigestUtils::sha256Hex;
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: This (digestSource) should be configurable

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+public class Digester {
+
+    private final DigestSource digestSource;
+
+    private final KeySource keySource;
+
+    private final String delimiter;
+
+    public Digester(KeySource keySource, String delimiter) {
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: Suggest adding a default constructor that associated KeySource with KeySources.random(some-sensible default)

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        this.keySource = keySource;
+        this.delimiter = delimiter;
+        this.digestSource = DigestUtils::sha256Hex;
+    }
+
+    private String saltValue(String value, String salt) {
+        return salt + delimiter + digestSource.digest(value);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: We need to digest the salt as well.

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public String digest(String value) throws Exception {
+        return saltValue(value, digestSource.digest(new String(keySource.getKey())));
+    }
+
+    public boolean validate(String value, String digest) {
+        String salt = StringUtils.substringBefore(digest, delimiter);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: What happens if there is no delimiter here? Ideally it would be an empty String, so it would also support the case of just a hashed value without salts. Maybe a test-case for this as well?

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/33/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/33/)

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        this.delimiter = delimiter;
+        this.digestSource = digestSource;
+    }
+
+    private String saltValue(String value, String salt) {
+        return digestSource.digest(salt) + delimiter + digestSource.digest(value);
+    }
+
+    public String digest(String value) throws Exception {
+        return saltValue(value, digestSource.digest(new String(keySource.getKey())));
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: Use new String(byte[], StandardCharsets.UTF_8) here, to avoid using a platform specific locale.

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    public Digester(KeySource keySource, String delimiter, DigestSource digestSource) {
+        this.keySource = keySource;
+        this.delimiter = delimiter;
+        this.digestSource = digestSource;
+    }
+
+    private String saltValue(String value, String salt) {
+        return digestSource.digest(salt) + delimiter + digestSource.digest(value);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: This should be: return salt + delimiter + digestSource.digest(salt + value);

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    public String digest(String value) throws Exception {
+        return saltValue(value, digestSource.digest(new String(keySource.getKey())));
+    }
+
+    public boolean validate(String value, String digest) {
+        String salt = StringUtils.substringBefore(digest, delimiter);
+        return (salt + delimiter + digestSource.digest(value)).equals(digest);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: return (salt + delimiter + digestSource.digest(salt + value)).equals(digest);

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+public class Digester {
+
+    private final DigestSource digestSource;
+
+    private final KeySource keySource;
+
+    private final String delimiter;
+
+    public Digester(DigestSource digestSource) {
+        this(KeySources.random(8), "", digestSource);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: Better to have 16 here instead of 8.

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package org.talend.daikon.crypto.digest;
+
+import org.apache.commons.codec.digest.DigestUtils;
+
+public class DigestSources {
+
+    public static DigestSource sha256() {
+        return DigestUtils::sha256Hex;
+    }
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: I'd prefer to use a standard algorithm for password hashing instead of sha256 actually, something like:

KeySpec spec = new PBEKeySpec(valueToDigest.toCharArray(), salt, 65536, 256);
SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
byte[] hash = factory.generateSecret(spec).getEncoded();

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/37/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/37/)

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java [hunk 11]
```diff
@@ -1,10 +1,51 @@
...
         return DigestUtils::sha256Hex;
     }
+
+    /**
+     * <p>
+     * Returns a PBKDF2 (with Hmac SHA256) digester. Please note <code>salt</code> must remain the same if you plan on
+     * having multiple {@link Digester} in your code.
+     * </p>
+     * <p>
+     * As recommandation, you may initialize a salt using {@link org.talend.daikon.crypto.KeySources#random(int)}.
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: "recommandation" -> "recommendation"

File: daikon-crypto/crypto-utils/src/test/java/org/talend/daikon/crypto/digest/DigesterTest.java [hunk 12]
```diff
@@ -14,7 +14,7 @@ public void shouldDigestValue() throws Exception {
         String value = "myPassword";
 
         // when
-        Digester digester = new Digester(KeySources.random(8), "-", DigestSources.sha256());
+        Digester digester = new Digester(KeySources.random(16), DigestSources.sha256());
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/test/java/org/talend/daikon/crypto/digest/DigesterTest.java]: Also add some tests for SHA-256 + an empty delimiter + a KeySource that supplies no salt. The output should be compared against a straighforward call to MessageDigest to check that it is ignoring the salt properly.

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 13]
```diff
@@ -31,15 +50,29 @@ public Digester(KeySource keySource, String delimiter, DigestSource digestSource
...
+    /**
+     * Digest a plain text value and returns a digested value.
+     *
+     * @param value The value to digest.
+     * @return A digested value using salt, delimiter and digested value.
+     * @throws Exception In any digest issue (depends on the implementation of {@link DigestSource} used).
+     */
     public String digest(String value) throws Exception {
-        return saltValue(value, digestSource.digest(new String(keySource.getKey())));
+        return saltValue(value, digestSource.digest(new String(keySource.getKey(), StandardCharsets.UTF_8)));
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: I don't think we should digest the salt here. Instead just convert it to Hex.

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java [hunk 14]
```diff
@@ -1,10 +1,51 @@
...
+ *
+ * @see Digester
+ */
 public class DigestSources {
 
+    /**
+     * @return A simple SHA256 digest for simplistic use cases (not recommended, see {@link #pbkDf2(byte[])}).
+     */
     public static DigestSource sha256() {
         return DigestUtils::sha256Hex;
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: I think here we should just return the raw bytes (+ so change the return type of DigestSource to byte[]). Then convert it to HEX in Digester. That way the PBKDF2 output is also encoded as hex.

File: daikon-crypto/crypto-utils/src/test/java/org/talend/daikon/crypto/digest/DigesterTest.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        final String digest = digester.digest(value);
+
+        // then
+        assertTrue(digester.validate("myPassword", digest));
+        assertFalse(digester.validate("MyPassword", digest));
+    }
+
+    @Test
+    public void shouldValidateValueWithDifferentDigester() throws Exception {
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/test/java/org/talend/daikon/crypto/digest/DigesterTest.java]: Also test here that using a different salt yields a different output.

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java (hunk 14 above)
Comment (fhuaulme) [reply daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: Given current API, the only place to encode digest to has (or base64) is done in one place (in DigestSource::pbkDf2). Changing DigestSource return to byte[] would force Digester class to encode bytes to base64 repeatedly (more times than it currently does).

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/38/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/38/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/39/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/39/)

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+     * having multiple {@link Digester} in your code.
+     * </p>
+     * <p>
+     * As recommendation, you may initialize a salt using {@link org.talend.daikon.crypto.KeySources#random(int)}.
+     * </p>
+     *
+     * @param salt The salt to be used to digest value.
+     * @return A {@link DigestSource} implementation using PBKDF2 and provided <code>salt</code>
+     */
+    public static DigestSource pbkDf2(byte[] salt) {
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: I think it's a bit confusing to take a salt here and also to take a salt from KeySources. Perhaps instead this method should not take a salt, and DigestSource should take a salt + digest parameter (rather than passing salt + digest) in Digester.

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/43/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/43/)

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+     * @return <code>true</code> if value matches digest, <code>false</code> otherwise.
+     */
+    public boolean validate(String value, String digest) {
+        if (delimiter == NO_DELIMITER) {
+            return (digestSource.digest(value, new byte[0])).equals(digest);
+        }
+        try {
+            final String saltBase64 = StringUtils.substringBefore(digest, valueOf(delimiter));
+            byte[] salt = decode(saltBase64.getBytes(EncodingUtils.ENCODING));
+            return (encode(salt) + delimiter + digestSource.digest(value, salt)).equals(digest);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: Can't we just use saltBase64 here instead of encode(salt)?

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @see Digester
+ */
+public class DigestSources {
+
+    /**
+     * @return A simple SHA256 digest for simplistic use cases (not recommended, see {@link #pbkDf2()}).
+     */
+    public static DigestSource sha256() {
+        return (data, salt) -> DigestUtils.sha256Hex(data);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: Could we just call DigestUtils.sha256 here + then use the encoder? Then it is consistently base64 instead of hex for one, and base64 for the other.

Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: I don't think we should ignore the salt here. Maybe we can check to see if salt is null or byte[0] + ignore it if so, otherwise add the salt in to the digest?

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/44/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/44/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/45/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/45/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/46/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/46/)

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java [hunk 19]
```diff
@@ -21,7 +22,17 @@
...
     public static DigestSource sha256() {
-        return (data, salt) -> DigestUtils.sha256Hex(data);
+        return (data, salt) -> {
+            final byte[] dataDigest = DigestUtils.sha256(data);
+            if (salt == null || salt.length == 0) {
+                return EncodingUtils.BASE64_ENCODER.apply(dataDigest);
+            } else {
+                byte[] result = new byte[salt.length + dataDigest.length];
+                System.arraycopy(salt, 0, result, 0, salt.length);
+                System.arraycopy(dataDigest, 0, result, salt.length, dataDigest.length);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: The salt isn't being digested here....we need to digest the salt as well and not just return the salt "as is".

Comment (fhuaulme) [reply daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: Very good point (sorry for missing this).

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * @return A simple SHA256 digest for simplistic use cases (not recommended, see {@link #pbkDf2()}).
+     */
+    public static DigestSource sha256() {
+        return (data, salt) -> {
+            final byte[] dataDigest = DigestUtils.sha256(data);
+            if (salt == null || salt.length == 0) {
+                return EncodingUtils.BASE64_ENCODER.apply(dataDigest);
+            } else {
+                byte[] result = new byte[salt.length + dataDigest.length];
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: Actually I think there's still a problem here, as we are using "salt.length" but actually copying the hashed salt. Ssame goes for the two lines below.

File: daikon-crypto/crypto-utils/src/test/java/org/talend/daikon/crypto/digest/DigesterTest.java [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+    public void shouldDigestValue() throws Exception {
+        // given
+        final String value = "myPassword";
+
+        // when
+        Digester digester = new Digester(KeySources.random(16), DigestSources.sha256());
+        final String digest = digester.digest(value);
+
+        // then
+        assertTrue(digest.contains("-"));
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/test/java/org/talend/daikon/crypto/digest/DigesterTest.java]: Could we add a sanity test here to check the length of the (decoded) salt matches 16?

File: daikon-crypto/crypto-utils/src/test/java/org/talend/daikon/crypto/digest/DigesterTest.java [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/test/java/org/talend/daikon/crypto/digest/DigesterTest.java]: Could we add a few more tests, something along the lines of:

TEst modifying salt portion of the value fails validation
Test modifying digest portion of the value fails validation
Test removing salt + delimiter part of the value fails digest fails validation

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/48/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/48/)

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+            if (salt == null || salt.length == 0) {
+                throw new IllegalArgumentException("Cannot use pbkDf2 with empty or null salt.");
+            }
+            try {
+                KeySpec spec = new PBEKeySpec(value.toCharArray(), salt, 65536, 256);
+                SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
+                final byte[] bytes = factory.generateSecret(spec).getEncoded();
+                return EncodingUtils.BASE64_ENCODER.apply(bytes);
+            } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                throw new IllegalStateException("Unable digest value.", e);
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/DigestSources.java]: "Unable to digest value"

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+import org.talend.daikon.crypto.EncodingUtils;
+import org.talend.daikon.crypto.KeySource;
+import org.talend.daikon.crypto.KeySources;
+
+import java.io.UnsupportedEncodingException;
+import java.util.stream.Collectors;
+
+import static java.lang.String.valueOf;
+import static java.util.Arrays.stream;
+import static java.util.stream.Collectors.joining;
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: This import isn't used anywhere

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+    private final DigestSource digestSource;
+
+    private final KeySource keySource;
+
+    private final char delimiter;
+
+    private static final int[] FORBIDDEN_DELIMITERS = { '/', '=' };
+
+    /**
+     * Creates a Digester using a {@link KeySources#random(int)} and "-" as delimiter for separating salt and digested
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: You could add that 16 is the default value used for the random bytes

File: daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+     * @param value The plain text value.
+     * @param digest The digest to compare with.
+     * @return <code>true</code> if value matches digest, <code>false</code> otherwise.
+     */
+    public boolean validate(String value, String digest) {
+        if (delimiter == NO_DELIMITER) {
+            return (digestSource.digest(value, new byte[0])).equals(digest);
+        }
+        try {
+            final String saltBase64 = StringUtils.substringBefore(digest, valueOf(delimiter));
```
Comment (coheigea) [inline daikon-crypto/crypto-utils/src/main/java/org/talend/daikon/crypto/digest/Digester.java]: Could we add in a check here that the value actually has the delimiter? If not then throw an exception.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/50/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/50/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/51/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/51/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/54/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/54/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/55/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/55/)
